File: ios/Classes/MapProjectionController.swift
import Foundation
import MapboxMaps
import UIKit

class MapProjectionController: Projection {
    func getMetersPerPixelAtLatitude(latitude: Double, zoom: Double) throws -> Double {
        return MapboxMaps.Projection.metersPerPoint(for: latitude, zoom: zoom)
    }

    func projectedMetersForCoordinate(coordinate: [String?: Any?]) throws -> ProjectedMeters {
        let projectedMeters = MapboxMaps.Projection.projectedMeters(for: convertDictionaryToCLLocationCoordinate2D(dict: coordinate)!)
        return ProjectedMeters(northing: projectedMeters.northing, easting: projectedMeters.easting)
    }

    func coordinateForProjectedMeters(projectedMeters: ProjectedMeters) throws -> [String?: Any?] {
        return MapboxMaps.Projection.coordinate(for: projectedMeters.toProjectedMeters()).toDict()
    }

    func project(coordinate: [String?: Any?], zoomScale: Double) throws -> MercatorCoordinate {
        let coordinate = MapboxMaps.Projection.project(convertDictionaryToCLLocationCoordinate2D(dict: coordinate)!, zoomScale: zoomScale)
        return coordinate.toFLTMercatorCoordinate()
    }

    func unproject(coordinate: MercatorCoordinate, zoomScale: Double) throws -> [String?: Any?] {
        return MapboxMaps.Projection.unproject(coordinate.toMercatorCoordinate(), zoomScale: zoomScale).toDict()
    }
}

File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/MapProjectionController.kt
package com.mapbox.maps.mapbox_maps

import com.mapbox.maps.MapboxMap
import com.mapbox.maps.mapbox_maps.pigeons.*

class MapProjectionController(private val mapboxMap: MapboxMap) : Projection {
  override fun getMetersPerPixelAtLatitude(latitude: Double, zoom: Double): Double {
    return mapboxMap.getMetersPerPixelAtLatitude(latitude, zoom)
  }

  override fun projectedMetersForCoordinate(coordinate: Map<String?, Any?>): ProjectedMeters {
    return mapboxMap.projectedMetersForCoordinate(coordinate.toPoint()).toFLTProjectedMeters()
  }

  override fun coordinateForProjectedMeters(projectedMeters: ProjectedMeters): Map<String?, Any?> {
    return mapboxMap.coordinateForProjectedMeters(projectedMeters.toProjectedMeters()).toMap()
  }

  override fun unproject(
    coordinate: MercatorCoordinate,
    zoomScale: Double
  ): Map<String?, Any?> {
    return mapboxMap.unproject(coordinate.toMercatorCoordinate(), zoomScale).toMap()
  }

  override fun project(
    coordinate: Map<String?, Any?>,
    zoomScale: Double
  ): MercatorCoordinate {
    return mapboxMap.project(coordinate.toPoint(), zoomScale).toFLTMercatorCoordinate()
  }
}